File: AddresDotNet/AddressBuilder.cs
using System;
using AddresDotNet.Enum;

namespace AddresDotNet;

/// <summary>
/// Строитель (Builder) для создания объекта адреса.
/// Позволяет пошагово задать компоненты адреса и затем собрать неизменяемый <see cref="Address"/>.
/// </summary>
public class AddressBuilder
{
    /// <summary>
    /// Регион адреса.
    /// </summary>
    private Region _region { get; set; }
    /// <summary>
    /// Населённый пункт.
    /// </summary>
    private Locality? _locality{ get; set; }
    /// <summary>
    /// Планировочный элемент.
    /// </summary>
    private PlanningElement? _planningElement{ get; set; }
    /// <summary>
    /// Улица или проезд.
    /// </summary>
    private Street? _street{ get; set; }
    /// <summary>
    /// Здание.
    /// </summary>
    private Building? _building{ get; set; }
    /// <summary>
    /// Помещение внутри здания.
    /// </summary>
    private Room? _room{ get; set; }
    /// <summary>
    /// Устанавливает регион адреса.
    /// </summary>
    /// <param name="name">Название региона.</param>
    /// <param name="type">Тип региона.</param>
    /// <returns>Текущий экземпляр <see cref="AddressBuilder"/> для цепочки вызовов.</returns>
    public AddressBuilder SetRegion(string name, RegionType type)
    {
        _region = new Region { Value = name, Type = type };
        return this;
    }
    /// <summary>
    /// Устанавливает населённый пункт.
    /// </summary>
    /// <param name="name">Название населённого пункта.</param>
    /// <param name="type">Тип населённого пункта.</param>
    /// <returns>Текущий экземпляр <see cref="AddressBuilder"/> для цепочки вызовов.</returns>
    public AddressBuilder SetLocality(string name, LocalityType type)
    {
        _locality = new Locality { Value = name, Type = type };
        return this;
    }
    /// <summary>
    /// Устанавливает планировочный элемент.
    /// </summary>
    /// <param name="name">Название планировочного элемента.</param>
    /// <param name="type">Тип планировочного элемента.</param>
    /// <returns>Текущий экземпляр <see cref="AddressBuilder"/> для цепочки вызовов.</returns>

    public AddressBuilder SetPlanningElement(string name, PlanningElementType type)
    {
        _planningElement = new PlanningElement { Value = name, Type = type };
        return this;
    }
    /// <summary>
    /// Устанавливает улицы.
    /// </summary>
    /// <param name="name">Название улицы.</param>
    /// <param name="type">Тип улицы.</param>
    /// <returns>Текущий экземпляр <see cref="AddressBuilder"/> для цепочки вызовов.</returns>
    public AddressBuilder SetStreet(string name, StreetType type)
    {
        _street = new Street { Value = name, Type = type };
        return this;
    }
    /// <summary>
    /// Устанавливает здание.
    /// </summary>
    /// <param name="number">Номер здания.</param>
    /// <param name="type">Тип здания.</param>
    /// <returns>Текущий экземпляр <see cref="AddressBuilder"/> для цепочки вызовов.</returns>
    public AddressBuilder SetBuilding(string number, BuildingType type)
    {
        _building = new Building { Value = number, Type = type };
        return this;
    }
    /// <summary>
    /// Устанавливает комнату/помещение в здании.
    /// </summary>
    /// <param name="number">Номер помещения.</param>
    /// <param name="type">Тип помещения.</param>
    /// <returns>Текущий экземпляр <see cref="AddressBuilder"/> для цепочки вызовов.</returns>
    public AddressBuilder SetRoom(string number, RoomType type)
    {
        _room = new Room { Value = number, Type = type };
        return this;
    }
    /// <summary>
    /// Собирает и возвращает объект адреса на основе заданных компонентов.
    /// </summary>
    /// <returns>Новый неизменяемый объект <see cref="Address"/>.</returns>
    /// <exception cref="InvalidOperationException">Если обязательный компонент (регион) не задан.</exception>
    public Address Build()
    {
        if (_region == null)
            throw new InvalidOperationException("Регион является обязательным компонентом адреса.");
        return new Address(_region,_locality,_planningElement, _street,_building,_room)
        {
            Region = _region,
            Locality = _locality,
            PlanningElement = _planningElement,
            Street = _street,
            Building = _building,
            Room =_room,
        };
    }
}

File: AddressKotlin/src/main/kotlin/builder/AddressBuilder.kt
package builder

import interfaces.IBuilder
import model.Address
import model.Building
import model.Locality
import model.PlanningElement
import model.Region
import model.Room
import model.Street

/**
 * Строитель для пошагового создания объекта [Address].
 * Реализует текучий интерфейс для объединения вызовов в цепочку.
 */

class AddressBuilder : IBuilder {

    private var _region: Region? = null
    private var _locality: Locality? = null
    private var _planningElement: PlanningElement? = null
    private var _street: Street? = null
    private var _building: Building? = null
    private var _room: Room? = null;


    /**
     * Устанавливает регион адреса.
     *
     * @param region Объект [Region].
     * @return Текущий экземпляр [AddressBuilder].
     */
    fun setRegion(region: Region): AddressBuilder = apply {
        _region = region
    }

    /**
     * Устанавливает населенный пункт.
     *
     * @param locality Объект [Locality].
     * @return Текущий экземпляр [AddressBuilder].
     */
    fun setLocality(locality: Locality): AddressBuilder = apply {
        _locality = locality
    }

    /**
     * Устанавливает элемент планировочной структуры (например, микрорайон).
     *
     * @param planningElement Объект [PlanningElement].
     * @return Текущий экземпляр [AddressBuilder].
     */
    fun setPlanning(planningElement: PlanningElement): AddressBuilder = apply {
        _planningElement = planningElement
    }

    /**
     * Устанавливает улицу.
     *
     * @param street Объект [Street].
     * @return Текущий экземпляр [AddressBuilder].
     */
    fun setStreet(street: Street): AddressBuilder = apply {
        _street = street
    }

    /**
     * Устанавливает здание или строение.
     *
     * @param building Объект [Building].
     * @return Текущий экземпляр [AddressBuilder].
     */
    fun setBuilding(building: Building): AddressBuilder = apply {
        _building = building;
    }

    /**
     * Устанавливает помещение (квартиру, офис).
     *
     * @param room Объект [Room].
     * @return Текущий экземпляр [AddressBuilder].
     */
    fun setRoom(room: Room): AddressBuilder = apply {
        _room = room
    }


    /**
     * Собирает и возвращает итоговый объект [Address].
     *
     * @return Сконструированный адрес.
     * @throws IllegalArgumentException Если не был установлен обязательный параметр [Region].
     */
    override fun build(): Address {
        val checkedRegion = requireNotNull(_region) { "Регион является обязательным параметром" }
        return Address(
            region = checkedRegion,
            locality = _locality,
            planningElement = _planningElement,
            street = _street,
            building = _building,
            room = _room
        )
    }
}